File: backend/app/models/analysis.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Literal
from uuid import uuid4

from pydantic import BaseModel, Field

PathStatus = Literal["running", "done", "error"]
PathKind = Literal["market", "single", "portfolio"]
Realm = Literal["a-share", "etf"]
ModelTier = Literal["primary", "local"]


class LlmRef(BaseModel):
    tier: ModelTier
    provider: str
    model: str


class ToolCall(BaseModel):
    id: str
    tool: str
    args: dict[str, Any] = Field(default_factory=dict)
    output: str | None = None
    output_ref: str | None = None
    status: Literal["ok", "error", "running"] = "ok"
    suggested_action: str | None = None


class AnalysisStep(BaseModel):
    id: str = Field(default_factory=lambda: uuid4().hex[:12])
    agent: str
    role: Literal["user", "assistant", "tool", "agent", "compact"] = "assistant"
    thought: str = ""
    result: str = ""
    tool_calls: list[ToolCall] = Field(default_factory=list)
    llm: LlmRef | None = None
    ts: str = Field(default_factory=lambda: datetime.now(UTC).isoformat())


class AnalyzeRequest(BaseModel):
    message: str
    session_id: str | None = None
    kind: PathKind = "single"
    realm: Realm = "a-share"
    target: str | None = None
    focus: str | None = None
    agent_ids: list[str] | None = None
    prompt_id: str | None = None
    primary_model: str | None = None
    enable_debate: bool = True
    debate_rounds: int = 1
    force_full: bool = False


class AnalysisPathIndexEntry(BaseModel):
    id: str
    title: str
    kind: PathKind
    realm: Realm
    status: PathStatus
    created: str
    updated: str
    target: str | None = None
    focus: str | None = None
    symbols: list[str] = Field(default_factory=list)
    judge_stance: str | None = None
    judge_one_liner: str | None = None
    data_as_of: str | None = None
    tags: list[str] = Field(default_factory=list)


class SseEvent(BaseModel):
    type: Literal["step", "token", "compaction", "phase", "error", "done", "reports", "memory"]
    step: AnalysisStep | None = None
    path_id: str | None = None
    step_id: str | None = None
    agent: str | None = None
    delta: str | None = None
    phase: str | None = None
    label: str | None = None
    snapshot: dict[str, Any] | None = None
    reports: dict[str, Any] | None = None
    message: str | None = None

File: frontend/src/types/analysis.ts
export type PageId = "chat" | "portfolio" | "workflow" | "context" | "agent-lib" | "ticket";
export type PathStatus = "running" | "done" | "error";
export type ModelTier = "primary" | "local";
export type AgentStatus = "active" | "draft";

export interface LlmRef {
  tier: ModelTier;
  provider: string;
  model: string;
}

export interface ToolCall {
  id: string;
  tool: string;
  args: Record<string, unknown>;
  output?: string | null;
  output_ref?: string | null;
  status: "ok" | "error" | "running";
  suggested_action?: string | null;
}

export interface AnalysisStep {
  id: string;
  agent: string;
  role: "user" | "assistant" | "tool" | "agent" | "compact";
  thought: string;
  result: string;
  tool_calls: ToolCall[];
  llm?: LlmRef | null;
  ts: string;
}

export interface AnalyzeRequest {
  message: string;
  session_id?: string | null;
  kind?: "market" | "single" | "portfolio";
  realm?: "a-share" | "etf";
  target?: string | null;
  focus?: string | null;
  agent_ids?: string[] | null;
  prompt_id?: string | null;
  primary_model?: string | null;
  enable_debate?: boolean;
  debate_rounds?: number;
  force_full?: boolean;
}

export interface ComposeRouteResult {
  kind: string;
  target: string | null;
  agent_ids: string[];
  prompt_id: string | null;
  enable_debate: boolean | null;
  rationale: string;
  memory_intent?: boolean;
}

export interface MemoryPreviewResult {
  status: string;
  memory_intent: boolean;
  symbol: string | null;
  hints: string[];
  summary: string;
}

export interface ContextSnapshot {
  summary: string;
  key_findings: { step_id: string; agent: string; text: string }[];
  carried_outputs: string[];
  total_raw_tokens: number;
  total_compressed_tokens: number;
}

export interface SseEvent {
  type: "step" | "token" | "compaction" | "phase" | "error" | "done" | "memory";
  step?: AnalysisStep;
  path_id?: string;
  step_id?: string;
  agent?: string;
  delta?: string;
  phase?: string;
  label?: string;
  snapshot?: ContextSnapshot;
  message?: string;
}

export interface PathIndexEntry {
  id: string;
  title: string;
  kind: string;
  realm: string;
  status: PathStatus;
  created: string;
  updated: string;
  target?: string | null;
  focus?: string | null;
  judge_stance?: string | null;
  judge_one_liner?: string | null;
  symbols?: string[];
}

export interface LlmHealthTier {
  ok: boolean;
  status: string;
  provider: string;
  model: string;
  label: string;
}

export interface LlmProviderInfo {
  id: string;
  tier: ModelTier;
  label: string;
  model: string;
  base_url: string;
}

export interface AgentSummary {
  id: string;
  name: string;
  model_tier: ModelTier;
  tools: string[];
  prompt_id: string;
  status?: AgentStatus;
  builtin?: boolean;
}

export interface ChatMessage {
  id: string;
  kind: "user" | "step";
  text?: string;
  step?: AnalysisStep;
}

export interface AnalysisMeta {
  kind: AnalyzeRequest["kind"];
  realm: AnalyzeRequest["realm"];
  focus: string | null;
  target: string | null;
  agent_ids?: string[] | null;
  prompt_id?: string | null;
  enable_debate?: boolean | null;
  route_rationale?: string | null;
}
